sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,445 @@
+        {
+            return; // nothing to log
+        }
+
+        // Convert binary content to bytes
+        using var memoryStream = new MaxLengthStream(_maxLength);
+        if (async)
+        {
+            await request.Content.WriteToAsync(memoryStream, cancellationToken).ConfigureAwait(false);
+        }
+        else
+        {
+            request.Content.WriteTo(memoryStream, cancellationToken);
+        }
+        var bytes = memoryStream.ToArray();
+
+        // Try to extract a text encoding from the headers
+        Encoding? requestTextEncoding = null;
+
+        if (request.Headers.TryGetValue("Content-Type", out var contentType) && contentType != null)
+        {
+            ContentTypeUtilities.TryGetTextEncoding(contentType, out requestTextEncoding);
+        }
+
+        // Log to event source
+        EventSourceSingleton.RequestContent(requestId, bytes, requestTextEncoding);
+    }
+
+    private void LogResponse(PipelineResponse response, string? responseId, double elapsed)
+    {
+        bool isError = response.IsError;
+        if (isError)
+        {
+            EventSourceSingleton.ErrorResponse(response, responseId, _sanitizer, elapsed);
+        }
+        else
+        {
+            EventSourceSingleton.Response(response, responseId, _sanitizer, elapsed);
+        }
+    }
+
+    private async Task LogResponseContent(PipelineResponse response, string? responseId, bool contentBuffered, bool async, CancellationToken cancellationToken)
+    {
+        var logErrorResponseContent = response.IsError && EventSourceSingleton.IsEnabled(EventLevel.Warning, EventKeywords.All);
+        var logResponseContent = EventSourceSingleton.IsEnabled(EventLevel.Informational, EventKeywords.All) || logErrorResponseContent;
+
+        if (!_logContent || !logResponseContent || response.ContentStream == null)
+        {
+            return;
+        }
+
+        // Try to extract a text encoding from the headers
+        Encoding? responseTextEncoding = null;
+
+        if (response.Headers.TryGetValue("Content-Type", out var contentType) && contentType != null)
+        {
+            ContentTypeUtilities.TryGetTextEncoding(contentType, out responseTextEncoding);
+        }
+
+        if (contentBuffered)
+        {
+            // Content is buffered, so log the first _maxLength bytes
+            ReadOnlyMemory<byte> contentAsMemory = response.Content.ToMemory();
+            var length = Math.Min(contentAsMemory.Length, _maxLength);
+            byte[] bytes = contentAsMemory.Span.Slice(0, length).ToArray();
+
+            LogFormattedResponseContent(response.IsError, responseId, bytes, responseTextEncoding);
+
+            return;
+        }
+
+        // We check content stream is not null above
+        if (response.ContentStream!.CanSeek)
+        {
+            using var memoryStream = new MaxLengthStream(_maxLength);
```

Comment:
I think using the MaxLengthStream wrapper is the easiest way to do this. Setting the capacity on a memory stream attempts to allocate that many bytes for the stream. So passing in int max value causes out of memory exceptions. The MaxLengthStream instead just limits how many bytes can be written.